### DISCUSSION TITLE
Variable as MQTT topic breaks HA integration

### DISCUSSION THREAD

**Discussion by septillion-git** [arendst/Tasmota#12710] [Ideas]
### PROBLEM DESCRIPTION
Two related problems of which the first breaks the HA integration:

Not al variables, for example `%hostname%`, are resolved in the discovery message for the Tasmota HA integration. Setting the MQTT topic to `%hostname%` does work for the actual MQTT topic used. Aka, if hostname is set to `Bedroom Light` and MQTT topic to `%hostname%` the actual MQTT topic is `[prefix]/Bedroom Light/`. But in the discovery message it's just published as `"t": "%hostname%"` which renders the device unavailable in HA. This did work with the old MQTT discovery.

Second I discovery when I tried to mitigate above issue is that I can use a space in the hostname but a space in the topic name is changed to an underscore. Even though a space in the topic works just fine, even with the Tasmota integration. If I set the topic to `%hostname%` which contains `Bedroom Light` and manually publish a new config in `tasmota/discovery` in which I set ``"t": "Bedroom Light"` it all works fine. So what's the reason for the no-spaces in topic names but do allow it in hostname?

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Sonoff Mini R2
- [x] Tasmota binary firmware version number used: 9.5.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: DIY Mode
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
21:07:46.039 MQT: Bedroom Light/stat/RESULT = {"NAME":"Sonoff MINIR2","GPIO":[32,0,0,0,160,0,0,0,224,576,0,0,0,0],"FLAG":0,"BASE":1}
21:07:46.249 MQT: Bedroom Light/stat/RESULT = {"Module":{"0":"Sonoff MINIR2"}}
21:07:46.503 MQT: Bedroom Light/stat/RESULT = {"GPIO0":{"32":"Button1"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"160":"Switch1"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"224":"Relay1"},"GPIO13":{"576":"LedLink_i"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
21:08:42.786 MQT: Bedroom Light/stat/STATUS = {"Status":{"Module":0,"DeviceName":"Bedroom Light","FriendlyName":["Bedroom Light"],"Topic":"%hostname%","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
21:08:42.818 MQT: Bedroom Light/stat/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin","RestartReason":"Software/System restart","Uptime":"0T00:09:20","StartupUTC":"2021-07-20T19:59:22","Sleep":50,"CfgHolder":4617,"BootCount":13,"BCResetTime":"2021-07-20T20:31:17","SaveCount":48,"SaveAddress":"F6000"}}
21:08:42.857 MQT: Bedroom Light/stat/STATUS2 = {"StatusFWR":{"Version":"9.5.0(tasmota)","BuildDateTime":"2021-06-17T08:26:35","Boot":7,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8285","CR":"369/699"}}
21:08:42.874 MQT: Bedroom Light/stat/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["UniFy NoT",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C8000100060000005A0A000000000000","00000080","00006000","00000000"]}}
21:08:42.907 MQT: Bedroom Light/stat/STATUS4 = {"StatusMEM":{"ProgramSize":607,"Free":396,"Heap":24,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"144051","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
21:08:42.942 MQT: Bedroom Light/stat/STATUS5 = {"StatusNET":{"Hostname":"Bedroom Light","IPAddress":"192.168.47.109","Gateway":"192.168.47.1","Subnetmask":"255.255.255.0","DNSServer":"1.1.1.1","Mac":"70:03:9F:5C:1C:D0","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
21:08:42.962 MQT: Bedroom Light/stat/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.47.20","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_5C1CD0","MqttUser":"mqtt","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
21:08:42.986 MQT: Bedroom Light/stat/STATUS7 = {"StatusTIM":{"UTC":"2021-07-20T20:08:42","Local":"2021-07-20T21:08:42","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+01:00","Sunrise":"05:08","Sunset":"20:43"}}
21:08:43.001 MQT: Bedroom Light/stat/STATUS10 = {"StatusSNS":{"Time":"2021-07-20T21:08:42","Switch1":"ON"}}
21:08:43.011 MQT: Bedroom Light/stat/STATUS11 = {"StatusSTS":{"Time":"2021-07-20T21:08:43","Uptime":"0T00:09:21","UptimeSec":561,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"UniFy NoT","BSSId":"B6:FB:E4:F7:E3:60","Channel":6,"Mode":"11n","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}}
```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE

- Setup device with a MQTT topic set to `%hostname%`
- Try to use in HA
- For second issue, use hostname with a space

### EXPECTED BEHAVIOUR
A device I can control from HA and is not unavailable aka the `%hostname%`-variable is also resolved to the value for the `tasmota/discovery`-topic.

### SCREENSHOTS
![image](https://user-images.githubusercontent.com/12911072/126389766-acd828d3-5d46-4aaa-a039-fb80ae8269ee.png)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
A good reason for no spaces in topic is that the syntax of various Tasmota commands, in particular Publish, does not support such topics. And IMO a space in host name is a syntax error, detected or not.

**Reply by septillion-git** [reply]
Yeah, I can live with an underscore instead of a space (although I like the space 👼). But greater issue is a) it can happen and b) it's not consistent looking at hostname.

**Reply by sfromis** [reply]
"It can happen" is not a valid argument. Tasmota does not try to enforce strict checking of everything able to cause an error. This is also a reason why you should just not be expecting perfect consistency. Tasmota is built to pack a ton of features into the very small space available on a tiny MCU. Do not expect that everything not rejected will work.

**Reply by septillion-git** [reply]
"It can happen" is not an argument in favor of supporting a space nor is it criticism nor do I expect perfect checking. But because it is done for the topic I only suggest to add the same policy to the hostname as well 😃 Yes, that makes that part a feature request. But with it baked in already for topic it can't be hard and makes things consistent. But more over, it was just a side issue to the main issue.

**Comment by septillion-git** [comment]
I don't really agree for this being a feature request... Forget the space-issue for now, that is completely separate. 

Setting the topic to a variable _**is**_ supported and the old MQTT discovery handled it correctly. So in my opinion it's just a bug the new discovery simply does not convert the variable.

**Comment by arendst** [comment]
Fixed in latest dev.
